src/Microsoft.AspNet.Proxy/ProxyMiddleware.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Proxy
+{
+    public class ProxyMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HttpClient _httpClient;
+        private ProxyOptions _options;
+
+        public ProxyMiddleware([NotNull] RequestDelegate next, [NotNull] ProxyOptions options)
+        {
+            _next = next;
+            if (string.IsNullOrEmpty(options.Host))
+            {
+                throw new InvalidOperationException("Options parameter must specify host");
+            }
+
+            // Setting default Port and Scheme if not specified
+            if (string.IsNullOrEmpty(options.Port))
+            {
+                options.Port = "80";
+            }
+
+            if (string.IsNullOrEmpty(options.Scheme))
+            {
+                options.Scheme = "http";
+            }
+
+            _options = options;
+            _httpClient = new HttpClient();
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            var requestMessage = new HttpRequestMessage();
+            if (string.Equals(context.Request.Method, "POST", StringComparison.OrdinalIgnoreCase))
+            {
+                var proxyStreamContent = new ProxyStreamContent(context.Request.Body);
+                requestMessage.Content = proxyStreamContent;
+                requestMessage.Method = new HttpMethod("POST");
+            }
+
+            // Copy the request headers
+            foreach (KeyValuePair<string, string[]> header in context.Request.Headers)
+            {
+                if (!requestMessage.Headers.TryAddWithoutValidation(header.Key, header.Value))
+                {
+                    bool success = requestMessage.Content.Headers.TryAddWithoutValidation(header.Key, header.Value);
```

Comment:
requestMessage.Content may be null. If it is, create a new ByteArrayContent(new byte[0]).